#Lecture: CS412 Machine Learning | Homework 4
####Instructor: Alper Özpınar
####Author: Mert Atabey Dincer
####Student ID: 20637
####Email: atabeymert@sabanciuniv.edu
####Date: 12.08.2020



In [9]:
import pandas as pd
import numpy as np
from sklearn.cluster import KMeans

In [10]:
url="https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"

columnNames = ["SepalLength", "SepalWidth", "PetalLength", "PetalWidth", "Output"]
df = pd.read_csv(url, index_col=None, header=None, names=columnNames) #Reads dataset from url

In [11]:
attributeCols = ["SepalLength", "SepalWidth", "PetalLength", "PetalWidth"]

X = df[attributeCols].to_numpy() #Seperate features from labels.
#We will be using only inputs since this is a unsupervised learning method.

In [12]:
from warnings import filterwarnings
filterwarnings('ignore')
numberOfClusters = [1,3,10,50,149]

for i in numberOfClusters:
  print("\n\nNUMBER OF CLUSTERS (n_clusters): {}".format(i))
  for j in range (1,11):
    kmeans = KMeans(n_clusters=i, n_init=j, random_state=5).fit(X)
    #print("\nNumber of runs (n_init): {}".format(j))                        #These prints have been used to figure out if n_init has any effect on output.
    #print("\nLABELS:\n{}".format(kmeans.labels_[:])                        #These helps to idenify eack n_init value's effect on results.
    #print("\nCLUSTER CENTERS:\n{}".format(kmeans.cluster_centers_[:]))
    #print("\nINERTIA: {:.3f}".format(kmeans.inertia_))
    #print("\n-------------------------------------------------------------------")
 
  print("\nLABELS:\n{}".format(kmeans.labels_[:]))   
  print("\nCLUSTER CENTERS:\n{}".format(kmeans.cluster_centers_[:]))
  print("\nINERTIA: {:.3f}".format(kmeans.inertia_))

  print("\n-------------------------------------------------------------------")



NUMBER OF CLUSTERS (n_clusters): 1

LABELS:
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0]

CLUSTER CENTERS:
[[5.84333333 3.054      3.75866667 1.19866667]]

INERTIA: 680.824

-------------------------------------------------------------------


NUMBER OF CLUSTERS (n_clusters): 3

LABELS:
[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 2 0 2 2 2 2 0 2 2 2 2
 2 2 0 0 2 2 2 2 0 2 0 2 0 2 2 0 0 2 2 2 2 2 0 2 2 2 2 0 2 2 2 0 2 2 2 0 2
 2 0]

CLUSTER CENTERS:
[[5.9016129  2.7483871  4.39354839 1.43387097]
 [5.006      3.418      1.464      0.244     ]
 [6.85       3.07368421 5.74210

#In terms of n_init: Determines number of times that k-means algorithm will be run. Normally, it uses different centroid seeds for each run and outputs best one in terms of inertia. In my case, for all runs inertia stayed same so I can say there was no difference for 1 or 10 times of running algorithm.


##Number of clusters: 1

* Number of runs (n_init): 1 INERTIA: 680.824

* Number of runs (n_init): 2 INERTIA: 680.824

* Number of runs (n_init): 3 INERTIA: 680.824

* Number of runs (n_init): 4 INERTIA: 680.824

* Number of runs (n_init): 5 INERTIA: 680.824

* Number of runs (n_init): 6 INERTIA: 680.824

* Number of runs (n_init): 7 INERTIA: 680.824

* Number of runs (n_init): 8 INERTIA: 680.824

* Number of runs (n_init): 9 INERTIA: 680.824

* Number of runs (n_init): 10 INERTIA: 680.824

-------------------------------------------------------------------


##Number of clusters: 3

* Number of runs (n_init): 1 INERTIA: 78.945

* Number of runs (n_init): 2 INERTIA: 78.945

* Number of runs (n_init): 3 INERTIA: 78.941

* Number of runs (n_init): 4 INERTIA: 78.941

* Number of runs (n_init): 5 INERTIA: 78.941

* Number of runs (n_init): 6 INERTIA: 78.941

* Number of runs (n_init): 7 INERTIA: 78.941

* Number of runs (n_init): 8 INERTIA: 78.941

* Number of runs (n_init): 9 INERTIA: 78.941

* Number of runs (n_init): 10 INERTIA: 78.941

-------------------------------------------------------------------


##Number of clusters: 10

* Number of runs (n_init): 1
INERTIA: 26.360

* Number of runs (n_init): 2
INERTIA: 26.360

* Number of runs (n_init): 3
INERTIA: 26.360

* Number of runs (n_init): 4
INERTIA: 26.360

* Number of runs (n_init): 5
INERTIA: 26.360

* Number of runs (n_init): 6
INERTIA: 26.360

* Number of runs (n_init): 7
INERTIA: 26.360

* Number of runs (n_init): 8
INERTIA: 26.360

* Number of runs (n_init): 9
INERTIA: 26.360

* Number of runs (n_init): 10
INERTIA: 26.360

-------------------------------------------------------------------


##Number of clusters: 50

* Number of runs (n_init): 1
INERTIA: 5.402

* Number of runs (n_init): 2
INERTIA: 5.402

* Number of runs (n_init): 3
INERTIA: 5.402

* Number of runs (n_init): 4
INERTIA: 5.402

* Number of runs (n_init): 5
INERTIA: 5.402

* Number of runs (n_init): 6
INERTIA: 5.402

* Number of runs (n_init): 7
INERTIA: 5.402

* Number of runs (n_init): 8
INERTIA: 5.402

* Number of runs (n_init): 9
INERTIA: 5.357

* Number of runs (n_init): 10
INERTIA: 5.357

-------------------------------------------------------------------


##Number of clusters: 149

* Number of runs (n_init): 1
INERTIA: 0.000

* Number of runs (n_init): 2
INERTIA: 0.000

* Number of runs (n_init): 3
INERTIA: 0.000

* Number of runs (n_init): 4
INERTIA: 0.000

* Number of runs (n_init): 5
INERTIA: 0.000

* Number of runs (n_init): 6
INERTIA: 0.000

* Number of runs (n_init): 7
INERTIA: 0.000

* Number of runs (n_init): 8
INERTIA: 0.000

* Number of runs (n_init): 9
INERTIA: 0.000

* Number of runs (n_init): 10
INERTIA: 0.000

#####In terms of n_clusters: Determines how many different classes the data will be divided into. In other words, how many labels or centroids will be created from out of given data.



#NUMBER OF CLUSTERS (n_clusters): 1



* In this case 1 label has been created. Since this is a 4 featured data, cluster center is consist of 4 coordinate and inertia is very high meaning that there is less or only one cluster. In other words, model is dealing with an underfitting (over-generalization) situation. At this point, it is hard do read and understand data. 

-------------------------------------------------------------------

#NUMBER OF CLUSTERS (n_clusters): 3
* While number of clusters is increasing, it can be seen that number of labels is increasing. 

-------------------------------------------------------------------

#NUMBER OF CLUSTERS (n_clusters): 10
* While number of clusters is increasing, inertia is decreasing. This is because of number of cluster centers is increasing in a particular area. So, every sample has a higher chance to find a closer cluster center than before.
This is why inertia is decreasing.

-------------------------------------------------------------------

#NUMBER OF CLUSTERS (n_clusters): 50
* Again number of labels and cluster centers is increasing while inertia is decreasing.

-------------------------------------------------------------------

#NUMBER OF CLUSTERS (n_clusters): 149
* Before filtering warnings I was getting this: /usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:7: ConvergenceWarning: Number of distinct clusters (147) found smaller than n_clusters (149). Possibly due to duplicate points in X. import sys

* Basically saying that there are idenical samples and k-means could not create distinct cluster centers for them. Even by this warning we can understand that It's been a long since we crossed the line of overfitting. Probably there will be no label/class for a new data point.
